### PR TITLE
Retrieve width and height of small icon using WINAPI to improve support of High-DPI displays

### PR DESCRIPTION
**Pull request by beru**
Hi, this PR improves High-DPI display support.

### CODE REVIEW

**Comment by csware** [comment]
Can you please provide before-after comparison screenshots?

**Comment by beru** [comment]
Please extract the attached archive file. I took screenshots of setting dialog.
Actually, its file format is 7z, not zip. so please rename it to `tgitss.7z` after downloading it.
[tgitss.zip](https://github.com/TortoiseGit/TortoiseGit/files/656074/tgitss.zip)

I shamelessly changed many part of the software suite (not only setting dialog property sheets) without scrutiny. In soviet russia, software test people.

**Comment by YueLinHo** [comment]
So, we don't need to re-create icon sets with these changes?

File: src/TortoiseIDiff/PicWindow.cpp [hunk 1]
```diff
@@ -1390,7 +1390,9 @@ bool CPicWindow::CreateButtons()
                                 NULL);
     if (hwndLeftBtn == INVALID_HANDLE_VALUE)
         return false;
-    hLeft = (HICON)LoadImage(hResource, MAKEINTRESOURCE(IDI_BACKWARD), IMAGE_ICON, 16, 16, LR_LOADTRANSPARENT);
+    int iconWidth = GetSystemMetrics(SM_CXSMICON);
+    int iconHeight = GetSystemMetrics(SM_CYSMICON);
+    hLeft = (HICON)LoadImage(hResource, MAKEINTRESOURCE(IDI_BACKWARD), IMAGE_ICON, iconWidth, iconHeight, LR_LOADTRANSPARENT);
```
Comment (YueLinHo) [inline src/TortoiseIDiff/PicWindow.cpp]: @csware Would it be great to have a helper? For encapsulation.

**Comment by YueLinHo** [comment]
cf. [Related issue on GitLab - Issue #2769: HiDPI support is not real great](https://gitlab.com/tortoisegit/tortoisegit/issues/2769)

**Comment by csware** [comment]
We don't need to re-create all icon sets, however, zooming icons might cause some less nice icons. I suppose that's better than icons which are way too small.

**Comment by csware** [comment]
@beru THanks for all the screenshots. One or two would have been enough :)

File: src/Resources/TortoiseProcENG.rc [hunk 2]
```diff
@@ -339,7 +339,7 @@ BEGIN
     GROUPBOX        "TortoiseGit",IDC_STATIC,7,7,286,92
     LTEXT           "&Language:",IDC_STATIC,14,20,86,8
     COMBOBOX        IDC_LANGUAGECOMBO,108,18,176,64,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
-    CHECKBOX        "Automatically check for &newer versions every week",IDC_CHECKNEWERVERSION,14,34,195,26,BS_VCENTER | BS_MULTILINE
+    CHECKBOX        "Automatically check for &newer versions every week",IDC_CHECKNEWERVERSION,14,36,195,18,BS_VCENTER
```
Comment (beru) [inline src/Resources/TortoiseProcENG.rc]: The reason I added this change is the checkbox's height is too tall and the control background overlaps to `Re-run First Start Wizard` button.

Please check below screenshot images.

ws000000
ws000001

File: src/TortoiseIDiff/PicWindow.cpp [hunk 3]
```diff
@@ -1468,10 +1470,12 @@ void CPicWindow::PositionChildren()
...
     {
-        SetWindowPos(hwndLeftBtn, HWND_TOP, rect.left+3, rect.top + HEADER_HEIGHT + (HEADER_HEIGHT-16)/2, 16, 16, SWP_FRAMECHANGED|SWP_SHOWWINDOW);
-        SetWindowPos(hwndRightBtn, HWND_TOP, rect.left+23, rect.top + HEADER_HEIGHT + (HEADER_HEIGHT-16)/2, 16, 16, SWP_FRAMECHANGED|SWP_SHOWWINDOW);
+        int iconWidth = GetSystemMetrics(SM_CXSMICON);
+        int iconHeight = GetSystemMetrics(SM_CYSMICON);
+        SetWindowPos(hwndLeftBtn, HWND_TOP, rect.left+iconWidth/4, rect.top + HEADER_HEIGHT + (HEADER_HEIGHT-iconHeight)/2, iconWidth, iconHeight, SWP_FRAMECHANGED|SWP_SHOWWINDOW);
+        SetWindowPos(hwndRightBtn, HWND_TOP, rect.left+iconWidth+iconWidth/2, rect.top + HEADER_HEIGHT + (HEADER_HEIGHT-iconHeight)/2, iconWidth, iconHeight, SWP_FRAMECHANGED|SWP_SHOWWINDOW);
         if (nFrames > 1)
-            SetWindowPos(hwndPlayBtn, HWND_TOP, rect.left+43, rect.top + HEADER_HEIGHT + (HEADER_HEIGHT-16)/2, 16, 16, SWP_FRAMECHANGED|SWP_SHOWWINDOW);
+            SetWindowPos(hwndPlayBtn, HWND_TOP, rect.left+iconWidth*2+iconWidth/2, rect.top + HEADER_HEIGHT + (HEADER_HEIGHT-iconHeight)/2, iconWidth, iconHeight, SWP_FRAMECHANGED|SWP_SHOWWINDOW);
```
Comment (beru) [inline src/TortoiseIDiff/PicWindow.cpp]: Added spaces with recent commit.

File: src/TortoiseProc/Settings/SetLookAndFeelPage.cpp [hunk 4]
```diff
@@ -32,23 +32,30 @@ extern MenuInfo menuInfo[];
...
+			mi.command != ShellSubMenu &&
+			mi.command != ShellSubMenuFile &&
+			mi.command != ShellSubMenuFolder &&
+			mi.command != ShellSubMenuLink &&
+			mi.command != ShellSubMenuMultiple) &&
+			(i == 0 || menuInfo[i - 1].menuID != mi.menuID))
 		{
-			HICON hIcon = reinterpret_cast<HICON>(::LoadImage(AfxGetResourceHandle(),
-					MAKEINTRESOURCE(menuInfo[i].iconID),IMAGE_ICON, 16, 16, LR_LOADTRANSPARENT ));
+			HICON hIcon = reinterpret_cast<HICON>(::LoadImage(
```
Comment (beru) [inline src/TortoiseProc/Settings/SetLookAndFeelPage.cpp]: You mean a helper function like this?

```
HICON LoadIconFromResource(WORD wInteger, bool isSmall, UINT fuLoad)
{
	int width = ::GetSystemMetrics(isSmall ? SM_CXSMICON : SM_CXICON);
	int height = ::GetSystemMetrics(isSmall ? SM_CYSMICON : SM_CYICON);
	return reinterpret_cast<HICON>(::LoadImage(AfxGetResourceHandle(),
					MAKEINTRESOURCE(wInteger),IMAGE_ICON, width, height, fuLoad));
}
```

By the way, I found interesting note about `LR_LOADTRANSPARENT` in this page.
https://msdn.microsoft.com/en-us/library/windows/desktop/ms648045(v=vs.85).aspx
> Do not use this option if you are loading a bitmap with a color depth greater than 8bpp.

**Comment by beru** [comment]
Enlarging small icons gives us poor visual experience because of low pixel density.
I wish there will be more larger icons in `src/Resources` folder.

File: src/Utils/TaskbarUUID.cpp [hunk 5]
```diff
@@ -145,11 +145,16 @@ void SetUUIDOverlayIcon( HWND hWnd )
...
 
+    int iconWidth = GetSystemMetrics(SM_CXSMICON);
+    int iconHeight = GetSystemMetrics(SM_CYSMICON);
+
     HICON icon = nullptr;
     if (!sicon.empty())
     {
         if (sicon.size() >= 4 && !_wcsicmp(sicon.substr(sicon.size() - 4).c_str(), L".ico"))
-            icon = (HICON)::LoadImage(nullptr, sicon.c_str(), IMAGE_ICON, 16, 16, LR_LOADFROMFILE | LR_SHARED);
+        {
```
Comment (beru) [inline src/Utils/TaskbarUUID.cpp]: Removed unnecessary braces in recent commit.

File: src/TortoiseProc/Settings/SetOverlayIcons.cpp [hunk 6]
```diff
@@ -132,8 +132,14 @@ void CSetOverlayIcons::ShowIconSet(bool bSmallIcons)
 	m_cIconList.DeleteAllItems();
 	m_ImageList.DeleteImageList();
 	m_ImageListBig.DeleteImageList();
-	m_ImageList.Create(16, 16, ILC_COLOR32 | ILC_MASK, 20, 10);
-	m_ImageListBig.Create(32, 32, ILC_COLOR32 | ILC_MASK, 20, 10);
+	int iconWidth = GetSystemMetrics(SM_CXSMICON);
+	int iconHeight = GetSystemMetrics(SM_CYSMICON);
+	m_ImageList.Create(iconWidth,
+					   iconHeight,
```
Comment (beru) [inline src/TortoiseProc/Settings/SetOverlayIcons.cpp]: Fixed it in recent commit.

File: src/Resources/TortoiseProcENG.rc (hunk 2 above)
Comment (csware) [inline src/Resources/TortoiseProcENG.rc]: Can you move it to another commit?

File: src/TortoiseProc/Settings/SetLookAndFeelPage.cpp (hunk 4 above)
Comment (csware) [inline src/TortoiseProc/Settings/SetLookAndFeelPage.cpp]: Yeah. I wouldn't use it everywhere but where it makes sense to make the code look a bit cleaner.

**Comment by csware** [comment]
> Enlarging small icons gives us poor visual experience because of low pixel density.

That's true. So what's better. To have small icons or have the bigger ones? THe issue is that we don't have any high quality originals for the icons. :(

**Comment by beru** [comment]
The strange thing is `gitconfig.ico` contains multiple sizes of images such as 48x48, 64x64, 256x256 but they are not somehow selected.

File: src/Resources/TortoiseProcENG.rc (hunk 2 above)
Comment (beru) [inline src/Resources/TortoiseProcENG.rc]: Sure, I can do that. Or should I make another PR for it?

Comment (csware) [inline src/Resources/TortoiseProcENG.rc]: Just add a separate commit. I'll cherry pick it.

File: src/TortoiseProc/Settings/SetLookAndFeelPage.cpp [hunk 7]
```diff
@@ -32,23 +32,30 @@ extern MenuInfo menuInfo[];
...
-		if ((menuInfo[i].command != ShellSeparator &&
-		   menuInfo[i].command != ShellSubMenu &&
-		   menuInfo[i].command != ShellSubMenuFile &&
-		   menuInfo[i].command != ShellSubMenuFolder &&
-		   menuInfo[i].command != ShellSubMenuLink &&
-		   menuInfo[i].command != ShellSubMenuMultiple) &&
-		   (i == 0 || menuInfo[i - 1].menuID != menuInfo[i].menuID))
+		const MenuInfo& mi = menuInfo[i];
+		if ((mi.command != ShellSeparator &&
+			mi.command != ShellSubMenu &&
```
Comment (csware) [inline src/TortoiseProc/Settings/SetLookAndFeelPage.cpp]: This can also be a separate commit with a message like "Use a temp variable instead of repeatedly calling the reference".

File: src/TortoiseProc/LogDlg.cpp [hunk 8]
```diff
@@ -364,8 +364,22 @@ BOOL CLogDlg::OnInitDialog()
...
 	m_JumpType.AddString(CString(MAKEINTRESOURCE(IDS_PROC_SELECTION_HISTORY)));
 	m_JumpType.SetCurSel(0);
-	m_JumpUp.SetIcon((HICON)::LoadImage(AfxGetInstanceHandle(), MAKEINTRESOURCE(IDI_JUMPUP), IMAGE_ICON, 16, 16, LR_DEFAULTCOLOR));
-	m_JumpDown.SetIcon((HICON)::LoadImage(AfxGetInstanceHandle(), MAKEINTRESOURCE(IDI_JUMPDOWN), IMAGE_ICON, 16, 16, LR_DEFAULTCOLOR));
+	int iconWidth = GetSystemMetrics(SM_CXSMICON);
+	int iconHeight = GetSystemMetrics(SM_CYSMICON);
+	m_JumpUp.SetIcon((HICON)::LoadImage(
+		AfxGetInstanceHandle(),
+		MAKEINTRESOURCE(IDI_JUMPUP),
+		IMAGE_ICON,
```
Comment (csware) [inline src/TortoiseProc/LogDlg.cpp]: Please keep in one line

File: src/TortoiseProc/Settings/SetLookAndFeelPage.cpp [hunk 9]
```diff
@@ -203,7 +205,9 @@ BOOL CSetLookAndFeelPage::OnInitDialog()
 
 	m_cMenuList.SetRedraw(false);
 
-	m_imgList.Create(16, 16, ILC_COLOR16 | ILC_MASK, 4, 1);
+	m_imgList.Create(GetSystemMetrics(SM_CXSMICON),
+					 GetSystemMetrics(SM_CYSMICON),
```
Comment (csware) [inline src/TortoiseProc/Settings/SetLookAndFeelPage.cpp]: Please keep on single line

File: src/TortoiseProc/Settings/SetLookAndFeelPage.cpp [hunk 10]
```diff
@@ -342,7 +346,9 @@ BOOL CSetExtMenu::OnInitDialog()
...
```
Comment (csware) [inline src/TortoiseProc/Settings/SetLookAndFeelPage.cpp]: Please keep on single line

File: src/Utils/MiscUI/HighColorTab.hpp [hunk 11]
```diff
@@ -45,7 +45,7 @@ namespace HighColorTab
 			return std::unique_ptr<CImageList>();
 		}
 
-		if( 0 == apILNew->Create( 16, 16, ILC_COLOR32|ILC_MASK, 0, 1 ) )
+		if( 0 == apILNew->Create( w, h, ILC_COLOR32|ILC_MASK, 0, 1 ) )
```
Comment (csware) [inline src/Utils/MiscUI/HighColorTab.hpp]: Please adjust spaces.

File: src/Utils/MiscUI/HighColorTab.hpp [hunk 12]
```diff
@@ -131,9 +131,9 @@ namespace HighColorTab
 
 	This method uses 32 bits image list creation default policy. */
 	template<typename TSheet>
-	bool UpdateImageList(TSheet& rSheet)
+	bool UpdateImageList(TSheet& rSheet, int w, int h)
 	{
-		return UpdateImageListFull<TSheet, HighColorTab::CHighColorListCreator>( rSheet );
+		return UpdateImageListFull<TSheet, HighColorTab::CHighColorListCreator>( rSheet, w, h );
```
Comment (csware) [inline src/Utils/MiscUI/HighColorTab.hpp]: Please adjust spaces.

File: src/Resources/TortoiseProcENG.rc (hunk 2 above)
Comment (beru) [inline src/Resources/TortoiseProcENG.rc]: Added a separated commit 3c2cd9a8dd3e6a51a4de388758e125bb6e673e9f.

File: src/TortoiseProc/LogDlg.cpp (hunk 8 above)
Comment (beru) [inline src/TortoiseProc/LogDlg.cpp]: Fixed coding style. https://github.com/TortoiseGit/TortoiseGit/pull/205/commits/3d8b61439ee4f2b5e4323a24d013d91f2f0f7618#diff-103cc511475bfe8e646cd59cf745b62c

File: src/TortoiseProc/Settings/SetLookAndFeelPage.cpp (hunk 9 above)
Comment (beru) [inline src/TortoiseProc/Settings/SetLookAndFeelPage.cpp]: Fixed coding style. https://github.com/TortoiseGit/TortoiseGit/pull/205/commits/3d8b61439ee4f2b5e4323a24d013d91f2f0f7618#diff-103cc511475bfe8e646cd59cf745b62c

File: src/TortoiseProc/Settings/SetLookAndFeelPage.cpp (hunk 10 above)
Comment (beru) [inline src/TortoiseProc/Settings/SetLookAndFeelPage.cpp]: Fixed coding style. https://github.com/TortoiseGit/TortoiseGit/pull/205/commits/3d8b61439ee4f2b5e4323a24d013d91f2f0f7618#diff-3712b53e9f8883544c78034214f13381

File: src/Utils/MiscUI/HighColorTab.hpp (hunk 11 above)
Comment (beru) [inline src/Utils/MiscUI/HighColorTab.hpp]: Adjusted space. Please check if it'd be alright in this way. https://github.com/TortoiseGit/TortoiseGit/pull/205/commits/3d8b61439ee4f2b5e4323a24d013d91f2f0f7618#diff-40ba2ca12f830d9b4547f8973a8d4b93

**Comment by csware** [comment]
Thank you very much!

Merged with ce4c0efe2a366745e4c5e2fd46316e889e4be036.

**Comment by csware** [comment]
@beru You seem to have High DPI enabled, I don't have and also don't have Win10. However, we have an [issue](https://gitlab.com/tortoisegit/tortoisegit/issues/2885) regarding High DPI issues with multiple monitors. Would be kewl if you could look into this!
